Add custom volumes for Distribution to add private CA certificates

**Is this a request for help?**:
No
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
FEATURE REQUEST

**Version of Helm and Kubernetes**:
Kubernetes: 1.14
Helm: 3.0.3

**Which chart**:
Distribution

**What happened**:
Cannot get MC to communicate with Distribution via HTTPS due to private CA certificate. HTTP works.

**What you expected to happen**:
Link the Distribution service with Mission Control. There is no option to mount a custom volume with the certificate as in the other Helm charts.

**How to reproduce it** (as minimally and precisely as possible):
1. Deploy MC and distribution with HTTPS (a similar issue may present with self signed certs)
2. Try to link Distribution using HTTPS

**Anything else we need to know**:
I was able to add my private CA certificates to MC/Artifactory(HA)/Xray by adding a custom volume to `customVolumes`/`customVolumeMounts` and setting up a `preStartCommand` as in other issues posted here. However, this option is not available for the Distribution chart, would be helpful if something similar was implemented.

Since the cert is trusted by MC but not by Distribution, my thoughts are that MC can reach Distribution but Distribution cannot reach back.

As a side note, it would be useful to implement this consistently across all the services (couldn't find `preStartCommand` for Xray server/indexer, had to get into the container(s) and run the `keytool` command manually).

Thanks in advance!

Hello @danielezer,
I have deployed this and I was able to mount my certificate into the Distribution container, but looking at the logs seems that `keytool` is not adding the certificate to `cacerts`. Here is the output from the command logs (from `kubectl logs` from the Distribution container):

```
Running custom Distribution preStartCommand command
Warning: use -cacerts option to access cacerts keystore
Certificate was added to keystore
keytool error: java.io.FileNotFoundException: /java/jdk-11.0.2+9/lib/security/cacerts (Permission denied)
```

I tried doing a `docker exec` as `root` into the container to run the command manually, but I got the same error. Taking a look at the folder, this is the `ls` output:
```bash
root@distribution-0:/java/jdk-11.0.2+9/lib/security# ls -la
total 356
drwxr-xr-x    2 500      500             98 Jan  1  1970 .
drwxr-xr-x    6 500      500           4096 Jan  1  1970 ..
-rw-r--r--    1 500      500           1253 Jan  1  1970 blacklisted.certs
-rw-r--r--    1 500      500         102883 Jan  1  1970 cacerts
-rw-r--r--    1 500      500           8815 Jan  1  1970 default.policy
-rw-r--r--    1 500      500         234959 Jan  1  1970 public_suffix_list.dat
``` 
From the dates, looks like the file wasn't modified. So the file exists but for some reason not even `root` can modify it. Have any clue what I'm missing?

I used the same command that I used for the other services, as suggested in your template:
```bash
/java/jdk-11.0.2+9/bin/keytool -importcert -keystore /java/jdk-11.0.2+9/lib/security/cacerts -storepass changeit -file <mount-location> -alias <alias> -noprompt
```

And logs show that SSL handshake didn't succeed:
```
2020-02-14 22:40:15,214 [http-nio-8080-exec-7] [ERROR] (o.j.b.d.s.p.t.EdgeNodesServiceImpl:104) Failed getting platform services from Mission Control
javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```

**Side note**
Does it make sense to also implement these in the `distributor` container? All our Artifactory nodes will have certificates issued from the same private CA. Asking because I noticed you guys also include a `preStartCommand` with the `keytool` example there, but I no see custom volume or mount options. Not sure if the `distributor` container interacts directly with the services. 🤔

Thanks in advance.

--- a/stable/distribution/CHANGELOG.md
+++ b/stable/distribution/CHANGELOG.md
@@ -1,6 +1,9 @@
 # JFrog Distribution Chart Changelog
 All changes to this project chart be documented in this file.
 
+## [3.6.0] - Feb 16, 2020
+* Add support for distributor `customVolumeMounts` and move `customVolumes` to `common`
+
 ## [3.5.3] - Feb 11, 2020
 * Add support for `preStartCommand`, `customVolumes` and `customVolumeMounts`
 

--- a/stable/distribution/Chart.yaml
+++ b/stable/distribution/Chart.yaml
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: distribution
 description: A Helm chart for JFrog Distribution
-version: 3.5.3
+version: 3.6.0
 appVersion: 1.8.0
 home: https://jfrog.com/platform/
 icon: https://raw.githubusercontent.com/jfrog/charts/master/stable/distribution/logo/distribution-logo.png

--- a/stable/distribution/README.md
+++ b/stable/distribution/README.md
@@ -178,14 +178,23 @@ distribution:
 There are cases where you'd like custom files mounted onto your container's file system.
 
 For this, there is a section for defining custom volumes in the [vaules.yaml](values.yaml).  By default they are left empty.
+You can mount custom volumes onto both the distribution and the distributor podsm like so:
 ```
-distribution:
+common:
   ## Add custom volumes
   customVolumes: |
   #  - name: custom-script
   #    configMap:
   #      name: custom-script
 
+distribution:
+  ## Add custom volumeMounts
+  customVolumeMounts: |
+  #  - name: custom-script
+  #    mountPath: "/scripts/script.sh"
+  #    subPath: script.sh
+
+distributor:
   ## Add custom volumeMounts
   customVolumeMounts: |
   #  - name: custom-script
@@ -238,6 +247,7 @@ The following table lists the configurable parameters of the distribution chart 
 | `logger.image.tag`                              | Tag for logger image                                                   | `1.30`                                                             |
 | `common.uid`                                    | Distribution and Distributor process user ID                           | `1020`                                                             |
 | `common.gid`                                    | Distribution and Distributor process group ID                          | `1020`                                                             |
+| `distribution.customVolumes`                    | Custom Volume Mounts for Distribution                                  | see [values.yaml](values.yaml)                                     |
 | `distribution.name`                             | Distribution name                                                      | `distribution`                                                     |
 | `distribution.image.pullPolicy`                 | Container pull policy                                                  | `IfNotPresent`                                                     |
 | `distribution.image.repository`                 | Container image                                                        | `docker.jfrog.io/jf-distribution`                                  |
@@ -246,7 +256,6 @@ The following table lists the configurable parameters of the distribution chart 
 | `distribution.service.loadBalancerSourceRanges` | Distribution service whitelist                                         | `[]`                                                               |
 | `distribution.customInitContainers`             | Custom init containers for Distribution                                |                                                                    |
 | `distribution.customVolumeMounts`               | Custom Volumes for Distribution                                        | see [values.yaml](values.yaml)                                     |
-| `distribution.customVolumes`                    | Custom Volume Mounts for Distribution                                  | see [values.yaml](values.yaml)                                     |
 | `distribution.externalPort`                     | Distribution service external port                                     | `80`                                                               |
 | `distribution.internalPort`                     | Distribution service internal port                                     | `8080`                                                             |
 | `distribution.masterKey`                        | Distribution Master Key (can be generated with `openssl rand -hex 32`) | `BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB` |
@@ -277,6 +286,7 @@ The following table lists the configurable parameters of the distribution chart 
 | `distributor.persistence.enabled`               | Distributor persistence volume enabled                                 | `true`                                                             |
 | `distributor.persistence.accessMode`            | Distributor persistence volume access mode                             | `ReadWriteOnce`                                                    |
 | `distributor.persistence.size`                  | Distributor persistence volume size                                    | `50Gi`                                                             |
+| `distributor.customVolumeMounts`                | Distributor Custom Volumes                                             | see [values.yaml](values.yaml)                                     |
 | `distributor.preStartCommand`                   | Distributor Command to run before the startup                          |                                                                    |
 | `distributor.nodeSelector`                      | Distributor node selector                                              | `{}`                                                               |
 | `distributor.affinity`                          | Distributor node affinity                                              | `{}`                                                               |

--- a/stable/distribution/templates/distribution-statefulset.yaml
+++ b/stable/distribution/templates/distribution-statefulset.yaml
@@ -221,7 +221,7 @@ spec:
         volumeMounts:
         - name: distributor-data
           mountPath: {{ .Values.distributor.persistence.mountPath | quote }}
-        {{- with .Values.distribution.customVolumeMounts }}
+        {{- with .Values.distributor.customVolumeMounts }}
 {{ tpl . $ | indent 8 }}
         {{- end }}
         resources:
@@ -308,7 +308,7 @@ spec:
 {{ toYaml . | indent 8 }}
     {{- end }}
       volumes:
-      {{- with .Values.distribution.customVolumes }}
+      {{- with .Values.common.customVolumes }}
 {{ tpl . $ | indent 6 }}
       {{- end }}
   {{- if not .Values.distribution.persistence.enabled }}

--- a/stable/distribution/templates/distribution-statefulset.yaml
+++ b/stable/distribution/templates/distribution-statefulset.yaml
@@ -221,6 +221,9 @@ spec:
         volumeMounts:
         - name: distributor-data
           mountPath: {{ .Values.distributor.persistence.mountPath | quote }}
+        {{- with .Values.distribution.customVolumeMounts }}
+{{ tpl . $ | indent 8 }}
+        {{- end }}
         resources:
 {{ toYaml .Values.distributor.resources | indent 10 }}
       - name: {{ template "redis.name" . }}

--- a/stable/distribution/values.yaml
+++ b/stable/distribution/values.yaml
@@ -19,16 +19,16 @@ rbac:
   role:
     ## Rules to create. It follows the role specification
     rules:
-    - apiGroups:
-      - ''
-      resources:
-      - services
-      - endpoints
-      - pods
-      verbs:
-      - get
-      - watch
-      - list
+      - apiGroups:
+          - ''
+        resources:
+          - services
+          - endpoints
+          - pods
+        verbs:
+          - get
+          - watch
+          - list
 
 networkpolicy:
   # Allows all ingress and egress
@@ -103,12 +103,12 @@ postgresql:
     size: 50Gi
     existingClaim:
   resources: {}
-#    requests:
-#      memory: "1Gi"
-#      cpu: "250m"
-#    limits:
-#      memory: "2Gi"
-#      cpu: "1"
+  #    requests:
+  #      memory: "1Gi"
+  #      cpu: "250m"
+  #    limits:
+  #      memory: "2Gi"
+  #      cpu: "1"
   nodeSelector: {}
   tolerations: []
   affinity: {}
@@ -173,6 +173,12 @@ common:
   uid: 1020
   gid: 1020
 
+  ## Add custom volumes
+  customVolumes: |
+  #  - name: custom-script
+  #    configMap:
+  #      name: custom-script
+
 logger:
   image:
     repository: 'busybox'
@@ -219,12 +225,6 @@ distribution:
   #    volumeMounts:
   #      - mountPath: "{{ .Values.distribution.persistence.mountPath }}"
   #        name: distribution-data
-
-  ## Add custom volumes
-  customVolumes: |
-  #  - name: custom-script
-  #    configMap:
-  #      name: custom-script
 
   ## Add custom volumeMounts
   customVolumeMounts: |
@@ -324,6 +324,12 @@ distributor:
     ##   GKE, AWS & OpenStack)
     ##
     # storageClass: "-"
+
+  ## Add custom volumeMounts
+  customVolumeMounts: |
+  #  - name: custom-script
+  #    mountPath: "/scripts/script.sh"
+  #    subPath: script.sh
 
   ## Custom command to run before Distribution startup.
   preStartCommand:

--- a/stable/distribution/values.yaml
+++ b/stable/distribution/values.yaml
@@ -19,16 +19,16 @@ rbac:
   role:
     ## Rules to create. It follows the role specification
     rules:
-      - apiGroups:
-          - ''
-        resources:
-          - services
-          - endpoints
-          - pods
-        verbs:
-          - get
-          - watch
-          - list
+    - apiGroups:
+      - ''
+      resources:
+      - services
+      - endpoints
+      - pods
+      verbs:
+      - get
+      - watch
+      - list
 
 networkpolicy:
   # Allows all ingress and egress
@@ -103,12 +103,12 @@ postgresql:
     size: 50Gi
     existingClaim:
   resources: {}
-  #    requests:
-  #      memory: "1Gi"
-  #      cpu: "250m"
-  #    limits:
-  #      memory: "2Gi"
-  #      cpu: "1"
+#    requests:
+#      memory: "1Gi"
+#      cpu: "250m"
+#    limits:
+#      memory: "2Gi"
+#      cpu: "1"
   nodeSelector: {}
   tolerations: []
   affinity: {}

